### PR TITLE
🐛 Synchonized plural generation with api-machinery

### PR DESCRIPTION
**Pull request by lauchokyip**
Fixes https://github.com/kubernetes-sigs/kubebuilder/issues/3402

### CODE REVIEW

**Review by camilamacedo86 — COMMENTED** [review]
/ok-to-test

**Comment by camilamacedo86** [comment]
@lauchokyip 

Could you please squash the commits ?

@varshaprasad96 @everettraven wdyt about this one?
Could you please give a hand on the review?

**Review by A-Hilaly — COMMENTED** [review]
+1 for trying to keep apimachinery and kubebuilder synced.

Quick question: this will probably break existing projects that already benefited from `gobuffalo/flect` pluralize mechanism. Maybe we should allow users to pass a set of "override" plural names (via flag, env/var...?), so at least they can work around this change?

**Comment by lauchokyip** [comment]
/hold 

Definitely feel like a breaking change

**Review by camilamacedo86 — COMMENTED** [review]
Hi @lauchokyip,

Could you please squash the commits?
Also, see that it is not passing in the lint you can run locally make lint locally to check it out. 

**By last, we need to provide a guidance for those that:**
- Scaffold the project with an previous version, create an api and then get a new kubebuilder version to create another API

It seems for me that would not work at all and this one is a breaking change. For other hand it seems a valid / reasonable request. I am just trying to think if it could be addressed for v4 or if we should only perform this change for an future version like go/v5.


c/c @everettraven @varshaprasad96 @Kavinjsir  wdyt about this one?

File: pkg/model/resource/utils.go
Comment (Kavinjsir) [inline pkg/model/resource/utils.go:71]: I think we can just leave a reference as well, which could be enough.
(This should also resolve the lint error as well)

image

**Review by Kavinjsir — COMMENTED** [review]
Thx for your contribution @lauchokyip ! Just give a first-round review and left a nit.

@camilamacedo86 
I'm Okay to make it consistent with the upstream kb plural system, though I personally find it hacky.  Basically, the upstream plural system looks far from 'perfect':
1. The k8s plural implementation itself is kind of broken, like `helmswomans` and `fishs`. Also, the implementation looks tricky, where it maintains a hardcoded `unpluralizedSuffixes`.
2. Like as you mentioned, if users are migrating their operator projects to the new kubebuilder with such an updated plural system. It may probably bring breaking change. (Though, it might not be a common scenario.)

Also in the future, we may probably change it once again, as such plural system might not be the 'ultimate' solution. (And then users might meet the similar scenario one more time.)

File: pkg/model/resource/utils.go
Comment (everettraven) [inline pkg/model/resource/utils.go:75]: Do we actually need this? `Endpoints` is a core Kubernetes resource and we likely aren't going to have to differentiate this in kubebuilder.

File: pkg/model/resource/utils.go
Comment (everettraven) [inline pkg/model/resource/utils.go:84]: We can probably remove this as per my previous comment

**Review by everettraven — CHANGES_REQUESTED** [review]
Thanks for the contribution @lauchokyip ! I have a couple suggested changes:

File: pkg/model/resource/utils.go
Comment (varshaprasad96) [inline pkg/model/resource/utils.go:90]: There is a piece of code in k8s/code-generator, that deals with this and some of the edge cases explicitly: https://github.com/kubernetes/gengo/blob/ab3349d207d4/namer/plural_namer.go#L36. May not be an immediate action item, but it would be preferable to have a TODO added here, to be able to resuse that piece of code. (It currently is very opinionated to spit out plurals for core k8s resources).

**Comment by lauchokyip** [comment]
> Hi @lauchokyip,
> 
> Could you please squash the commits? Also, see that it is not passing in the lint you can run locally make lint locally to check it out.
> 
> **By last, we need to provide a guidance for those that:**
> 
> * Scaffold the project with an previous version, create an api and then get a new kubebuilder version to create another API
>   Could you please add the steps in the description of this PR? What guidance/steps should we provide in this case?
> 
> It seems for me that would not work at all and this one is a breaking change. For other hand it seems a valid / reasonable request. I am just trying to think if it could be addressed for v4 or if we should only perform this change for an future version like go/v5.
> 
> c/c @everettraven @varshaprasad96 @Kavinjsir wdyt about this one?

Hi @camilamacedo86 , do you providing the migration documents? Any guidance on how I could do that?  I don't know the code well enough to answer this question - if the user generate a CRD named Helmswoman and the plural is set to `Helmswomen`, and if the user were to generate another CRD for different version `v2`, kind `Helmswoman`, would the plural be set to `Helmswomans` ? If the plural were to set to `Helmswomans`, will that break anything besides the generated file name?

**Review by everettraven — APPROVED** [review]
Code looks good to me, thanks @lauchokyip !

**Comment by camilamacedo86** [comment]
Hi @everettraven @varshaprasad96 @Kavinjsir @lauchokyip,

We have encountered several deprecations, and it's essential that we deprecate the declarative plugin as soon as possible. You can find the details in this pull request: https://github.com/kubernetes-sigs/kubebuilder/pull/3395

To ensure a cleaner and more maintainable codebase as we continue to grow, we plan to remove all deprecations at the beginning of next year, leading to a major bump for Kubebuilder 4.0.0.

However, since this change will have a significant impact on projects that are already scaffolded, I suggest we consider the following:

- Let's carefully evaluate whether this change is necessary, taking into account the comments shared by @Kavinjsir in this pull request: https://github.com/kubernetes-sigs/kubebuilder/pull/3433#pullrequestreview-1453231505.
- In the pull request, it would be helpful to document the manual steps required to update existing projects once this change is merged. @lauchokyip, could you please include these steps in the PR description? We need to provide guidance on how to modify projects built prior to this change, ensuring they won't be affected when a new version of Kubebuilder is released. Please specify the necessary file modifications and steps to be taken. (It is essential because it will help us properly document it in the release and migration guides)
- @lauchokyip, please ensure that the new code is only used for go/v4 and doesn't impact go/v3, which is already deprecated. By doing so, when we eventually bump to 4.x, the condition will keep the code as the default. 
- Then, I think we might want to consider whether to keep this PR for future go/v5 changes. Once @lauchokyip implements the changes for impact only go/v4 as described above, we can modify the PR for go/v5 easily and merge it at the appropriate time.

Please share your thoughts on these suggestions.

**Comment by lauchokyip** [comment]
> @lauchokyip We are interested in resolving this issue. Could you please rebase this so that we revisit it again. Thanks for taking the time for working on this.

Let me get back to this in a bit

**Comment by lauchokyip** [comment]
HI @varshaprasad96 , `make generate` are failing, can you please check it? I even `git clone` a brand new repo and it still failed
```
TODO: update tutorial
ERRO[0009] error fixing cronjob_webhook.go: unable to find the content to be replaced 
exit status 1
make: *** [Makefile:76: generate-docs] Error 1
```

**Comment by Kavinjsir** [comment]
> HI @varshaprasad96 , `make generate` are failing, can you please check it? I even `git clone` a brand new repo and it still failed
> 
> ```
> TODO: update tutorial
> ERRO[0009] error fixing cronjob_webhook.go: unable to find the content to be replaced 
> exit status 1
> make: *** [Makefile:76: generate-docs] Error 1
> ```
> 
> Does `make generate` work for you?

@lauchokyip Would you rebase the branch based off the latest master branch? There might be an issue for `make generate` previously. The current master branch should have it fixed. Could you try and see if it works? Feel free to let me know if further questions rasied.

**Comment by camilamacedo86** [comment]
Closed based on:
- Huge breaking change https://github.com/kubernetes-sigs/kubebuilder/pull/3433#issuecomment-1585538375 without a good justification. Why the change is not proposed to apimachinery?
- https://github.com/kubernetes-sigs/kubebuilder/issues/3402#issuecomment-2121490772
- https://github.com/kubernetes-sigs/kubebuilder/issues/3402#issuecomment-2122171734

**Pull request closed without merging**
